Retrieve real account status

Mayisha opened the pull request:
Fixes #1822

### Description
Retrieving the real account data from Stripe API and showing them in the UI.

![email exists](https://user-images.githubusercontent.com/33387139/133590914-27b5578c-df37-422c-8dda-851bd224353f.png)

### Changes proposed
- Created a 'WC_Stripe_Account' class to manage account data
- Ported some functions from `WCPay` here
- Using WP Option for caching account data. Similar caching is also implemented in WCPay and useful for handling Stripe rate limit. Initially implemented without the caching and after a few requests started getting `cURL error 28: Resolving timed out after 10000 milliseconds`.  
- Added a global variable `wc_stripe_admin_settings` intended to include all account settings related data.

### Testing instructions
With test keys
- Add test key from the old UI
- Enable the `_wcstripe_feature_upe_settings` feature flag to check the account status in the new UI

With live keys
- Add live key from the old UI
- Enable the `_wcstripe_feature_upe_settings` feature flag to check the account status in the new UI

### QA steps
To be added.....

### Pending
- What to show when Email is empty. (The email property is not present in response when requested using test key)
- What to show when error occurs and failed to retrieve account data

Mayisha committed:
Create 'WC_Stripe_Account' class

Mayisha committed:
Initialize 'WC_Stripe_Account' on load to cache account data

Mayisha committed:
Add global variable 'wc_stripe_admin_settings' and use accountSatus from it

Mayisha commented:
Will add some tests after getting a review on the implementation and answers to the pending questions.

tommyshellberg commented on includes/class-wc-stripe-account.php:
```diff
@@ -0,0 +1,160 @@
+	const ACCOUNT_API    = 'account';
+
+	public function __construct() {
+		self::cache_account();
```
Would we want to try reading the cached data first before doing the API request?
This reads to me like when the constructor fires it will do the API request and cache it but we'd want to check if the cached data exists first?

Mayisha replied:
That's a valid concern. :+1:  
I will rearrange the logic or if I just call the `get_cached_account_data` here instead of `cache_account()` that should also work. Because all the checks are there sequentially.

tommyshellberg replied:
Yep, calling `get_cached_account_data()` looks right to me.

tommyshellberg commented:
My thoughts on the pending questions:

> What to show when Email is empty. (The email property is not present in response when requested using test key).

I think you could either just use a fallback string like 'test account' here:

https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1869/files#diff-4314dc04e3c11401b896736a579e7b2cce4634bab67b790ef2872cc6b39edfb3R154

Or, in the React app, if the email is an empty string, render a pill in the header which says, `Test Mode Enabled` which I believe was a recommendation from Jarek.

> What to show when error occurs and failed to retrieve account data

The `AccountDetails` component has a check for `accountStatus.error` and [shows an error message](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/25ad12471502ad5ce013aebb59a98484fb2aaac7/client/settings/account-details/index.js#L56). So if you fail to get an account object you could add the `error` prop with a `true` value to `accountStatus`.

Since I'm going AFK I'll un-assign myself but I hope this was helpful. :)

frosso commented on includes/admin/class-wc-stripe-settings-controller.php:
```diff
@@ -81,6 +81,11 @@ public function admin_scripts() {
+		$settings = [
+			'accountStatus' => WC_Stripe_Account::get_account_status(),
+		];
+		wp_localize_script( 'woocommerce_stripe_admin', 'wc_stripe_admin_settings', $settings );
```
Out of curiosity - the `woocommerce_stripe_admin` is already part of `wp_localize_script` a few lines above - why add a new global `wc_stripe_admin_settings`, could we have leveraged the existing `wc_stripe_settings_params`?

Mayisha replied:
I was thinking of keeping the saved settings related values in one variable ( `wc_stripe_admin_settings` ) and the other params (flags and all) in the previous one. We can have all in the same variable though. Does not make much differences.

frosso commented on includes/admin/class-wc-stripe-settings-controller.php:
```diff
@@ -81,6 +81,11 @@ public function admin_scripts() {
 		wp_localize_script( 'woocommerce_stripe_admin', 'wc_stripe_settings_params', $params );
 
+		$settings = [
+			'accountStatus' => WC_Stripe_Account::get_account_status(),
```
`WC_Stripe_Account` seems to be a singleton in the app - static methods are pretty hard to test.
What are your thoughts about reducing the number of static methods, and having `WC_Stripe_Account` injected in the constructor of `WC_Stripe_Settings_Controller`? That way things will be a bit more testable.

frosso commented on client/settings/payment-settings/index.js:
```diff
@@ -101,24 +102,19 @@ const AccountSettingsDropdownMenu = () => {
 			<CardHeader className="account-details__header">
 				<h4 className="account-details__header">
-					{ accountStatusMock.email }
+					{ accountStatus.email }
```
Like Thomas mentioned - we'll need the `test mode` pill in test mode.

frosso reviewed (commented):
Taking over the review, since Thomas is AFK - I think he addressed the questions, I gave this an initial look and it seems like we're on the right track!
Add a few tests, the error message, the `test mode` pill and we're pretty much good to go!

frosso commented on includes/class-wc-stripe-account.php:
```diff
@@ -0,0 +1,160 @@
+	/**
+	 * Caches account data for a period of time.
+	 */
+	private static function cache_account() {
```
When the merchant switches "test mode", I'm guessing that this cache will need to be invalidated. Is there something that does it?
Maybe the "test mode" can be part of the cache key.

Mayisha replied:
Updated the code. If the current mode does not match the cached mode, fetching the data again.

frosso commented on includes/class-wc-stripe-account.php:
```diff
@@ -0,0 +1,160 @@
+		$account_cache['expires'] = time() + $expiration;
+
+		// Create or update the account option cache.
+		if ( false === get_option( self::ACCOUNT_OPTION ) ) {
+			add_option( self::ACCOUNT_OPTION, $account_cache, '', 'no' );
+		} else {
+			update_option( self::ACCOUNT_OPTION, $account_cache, 'no' );
+		}
```
Why not just use `update_option`? Is it necessary to check if the option was set before?

BTW, WordPress has a similar mechanism for this "caching" behavior - transients: https://developer.wordpress.org/apis/handbook/transients/#using-transients
You can set a cache and you'll be able to cache the response for a certain amount of time. When the cache expires the data is removed, tho.

Mayisha replied:
Thanks for the link. Using transient now. :smile:

frosso replied:
Something worth mentioning - the transient data "expires", not sure if that was your intent as well

Mayisha committed:
Refactor code
- Use `Test Mode` pill
- Avoid declaring static method
- Use `wc_stripe_settings_params`
- Refresh cache when current mode does not match the cached mode

Mayisha committed:
Add test

frosso commented on client/settings/payment-settings/index.js:
```diff
@@ -1,3 +1,7 @@
+/* global wc_stripe_settings_params */
+/**
+ * External dependencies
+ */
```
We won't need to add these comments anymore 👍 
```suggestion
```

frosso commented on includes/admin/class-wc-stripe-settings-controller.php:
```diff
@@ -10,7 +10,15 @@
+	private $account;
+
 	public function __construct() {
+		$this->account = new WC_Stripe_Account();
```
What do you think about injecting this in the constructor, rather than instantiating it here?
That way you can more easily mock it for the tests

frosso reviewed (commented):
Let's try the approach we discussed over PM for the tests on `WC_Stripe_Account_Test`

frosso commented on includes/class-wc-stripe-account.php:
```diff
@@ -0,0 +1,159 @@
+			return $account;
+		}
+
+		$account = $this->cache_account();
+		return $account;
```
```suggestion
		return $this->cache_account();
```

frosso commented on includes/class-wc-stripe-account.php:
```diff
@@ -0,0 +1,159 @@
+	 *
+	 * @return array Account data or empty if failed to retrieve account data.
+	 */
+	public function get_cached_account_data() {
```
What are your thoughts on having 2 separate transient keys: one for "live" and the other for "test" mode.

That way you don't need to add the checks on the `read_account_from_cache` or `cache_account` methods - the transient key could be "generated" in a method in this class (e.g.: `get_transient_key()`) for both `read_account_from_cache` and `cache_account`.

Mayisha committed:
Inject account in constructor of WC_Stripe_Settings_Controller

Mayisha committed:
Inject $connect in WC_Stripe_Account

Mayisha committed:
Use separate transients for test mode and live mode

Mayisha committed:
Inject stripe api in WC_Stripe_Account constructor

Mayisha committed:
Update test
- Create a new helper class WC_Helper_Stripe_Api to mock the static methods of WC_Stripe_API
- Mock WC_Stripe_Connect

Mayisha committed:
remove declaration of External dependencies

Mayisha committed:
add file path

Mayisha committed:
use 'call_user_func' for compatibility with php < 5.2

frosso commented on includes/admin/class-wc-stripe-settings-controller.php:
```diff
@@ -10,7 +10,13 @@
  */
 class WC_Stripe_Settings_Controller {
-	public function __construct() {
+	/**
+	 * Constructor
+	 *
+	 * @param WC_Stripe_Account $account Stripe account
+	 */
+	public function __construct( WC_Stripe_Account $account ) {
+		$this->account = $account;
```
It's probably best to also declare the property on the class' definition (also valid for the `WC_Stripe_Account`, `WC_Stripe` and `WC_Stripe_Account_Test` classes)
```suggestion
	/**
	 * The Stripe account instance.
	 *
	 * @var WC_Stripe_Account
	 */
	private $account;

	/**
	 * Constructor
	 *
	 * @param WC_Stripe_Account $account Stripe account
	 */
	public function __construct( WC_Stripe_Account $account ) {
		$this->account = $account;
```

frosso commented on includes/class-wc-stripe-account.php:
```diff
@@ -0,0 +1,166 @@
+
+	const LIVE_ACCOUNT_OPTION = 'wcstripe_account_data_live';
+	const TEST_ACCOUNT_OPTION = 'wcstripe_account_data_test';
+	const ACCOUNT_API         = 'account';
```
It looks like `ACCOUNT_API` is not used anywhere in the codebase
```suggestion
```

frosso commented on includes/class-wc-stripe-account.php:
```diff
@@ -0,0 +1,166 @@
+	 * @return array Account status data or empty if failed to retrieve account data.
+	 */
+	public function get_account_status() {
+		$account          = json_decode( json_encode( $this->get_cached_account_data() ), true );
```
(minor) Is the `json_decode`/`json_encode` combo happening here to convert a possible object to an associative array?

The calls _could_ be a little bit expensive because of the casting being done and depending on the amount of data being decoded/encoded.

I see that `get_cached_account_data()` calls either `read_account_from_cache()` or `cache_account()`.
`read_account_from_cache()` already calls the combo of `json_decode`/`json_encode`. So in that case it would be a superfluous call.
The other instance is `cache_account()`.

Maybe we could put the `json_decode`/`json_encode` combo in `cache_account()` and remove it from `get_account_status()`, so that we can avoid doing a possibly unnecessary call to `json_decode`/`json_encode`.

frosso reviewed (approved):
The changes look good, good job with the tests 💯 

I left a comment on ensuring the properties are declared on the classes, when added.
The rest is pretty minor.

Mayisha committed:
Address feedbacks
- declare private variables
- remove unnecessary const
- move json_encode/json_decode inside 'cache_account()'

Mayisha commented:
Thanks for the feedback. Addressed them :tada:

Mayisha committed:
Merge branch 'develop' into add/1822-real-account-status

frosso reviewed (approved):
Still looking good, thanks for taking the time! :shipit:

Mayisha committed:
Merge branch 'develop' into add/1822-real-account-status
